---
title: "Best local alignment of two nucleotide sequences according to Smith-Waterman algorithm with smithWaterman"
author: "Emanuele Cavalleri"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Best local alignment of two nucleotide sequences according to Smith-Waterman algorithm with smithWaterman}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, include = FALSE}
library(BiocStyle)
library(knitr)
```

```{r, message=FALSE}
library(Biostrings)
library(benchmarkme)
library(ShortRead)

library(smithWaterman)
```

# Introduction

The **Smith–Waterman** algorithm performs local sequence alignment; that is, for determining similar regions between two strings of nucleic acid sequences or protein sequences ([Wikipedia entry](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm)). Instead of looking at the entire sequence, the Smith–Waterman algorithm compares segments of all possible lengths and optimizes the similarity measure.

The `r Rpackage("smithWaterman")` package provides functions to apply the Smith-Waterman algorithm on nucleotide sequences and to show its result.

# Smith-Waterman algorithm

## Applying **Smith-Waterman** algorithm to strings and showing its result

Applying **Smith-Waterman** algorithm to two strings is easy as shown in this example:

```{r}
X <- "TATCT"
Y <- "TAGTCT"
match <- 7
mismatch <- -10
gap <- -7
# Same with: mismatch <- 10 and/or gap <- 7

sw <- smith_waterman_string(X, Y, match, mismatch, gap)
sw
```

*show()* method allows us to better visualize the results.

```{r}
sw$show()
```

## Applying **Smith-Waterman** algorithm to DNAString objects

Data type **DNAString** (S4 class) is a container for storing a (potentially very long) individual DNA sequence, with efficient methods for access and manipulation.

We can apply **Smith-Waterman** algorithm to two DNAString objects as shown in this example:

```{r}
X <- DNAString(X)
Y <- DNAString(Y)

sw <- smith_waterman_DNAString(X, Y, match, mismatch, gap)
sw$show()
```

A real-case example based on *phiX174Phage*'s reads to also test the efficiency of the algorithm in terms of time:

```{r}
data(phiX174Phage)
X <- subseq(phiX174Phage[1], 1, 500) # First 500 nucleotides
Y <- subseq(phiX174Phage[2], 1, 500)

start_time <- Sys.time() # Start the stopwatch
# Execute it for ten times
for (i in 1:10)
  sw <- smith_waterman_DNAString(X, Y, match, mismatch, gap)
end_time <- Sys.time() # Stop the stopwatch

cat("Best local alignment's score according to Smith-Waterman algorithm:",
    sw$get_alignment_score())
```

```{r out.width="20%", echo = FALSE}
url <- "https://www.humboldtmfg.com/product-originals/_lrg@1x/H-2260.png"
knitr::include_graphics(url)
```

```{r}
cat("Mean execution time", round(end_time - start_time, 1) / 10, "second(s) on a(n)",
    get_cpu()$model_name, "having", round(get_ram()/1000000000, 2), "GB of RAM")
```

Another real-case example with nucleotide sequences coming from a **fastq** file:

```{r}
fastqfile <- system.file("extdata", "SP1.fq", package="smithWaterman")
seqset <- readDNAStringSet(fastqfile, format="fastq", with.qualities=TRUE)
X <-  seqset[1]
Y <-  seqset[246]

sw <- smith_waterman_DNAString(X, Y, match, mismatch, gap)
cat("Best local alignment's score according to Smith-Waterman algorithm:",
    sw$alignment_score)
```

A last real-case example based on *HNF4alpha*'s sequences:

```{r}
data(HNF4alpha)
X <-  HNF4alpha[1]
Y <-  HNF4alpha[6]

sw <- smith_waterman_DNAString(X, Y, match, mismatch, gap)
sw$show()
```

## Applying **Smith-Waterman** algorithm to string and/or DNAString objects

*smith_waterman* function permits you to bypass the problem associated with different type of objects as input. This function is the one suggested to locally align two nucleotide sequences in any format (i.e.: strings or DNAString objects).

```{r}
sw <- smith_waterman(HNF4alpha[1], "AGGCCAAAGTCCT", match, mismatch, gap)
sw$show()
```

## SessionInfo
```{r}
sessionInfo()
```